### PR TITLE
Add content_filter hook

### PR DESCRIPTION
**Pull request by MrPetovan**
See: https://friendica.mrpetovan.com/display/735a2029715ab80d1874cf6808556656

This hook collects all the potential filtering reasons for a single post and allows to display them all instead of having multiple collapsed posts.

Additionally, there is now a setting to disable the content warning set by authors on Mastodon and Pleroma.

It was an opportunity to update the Addons documentation dating from 2012!

Addon PR: https://github.com/friendica/friendica-addons/pull/566

### CODE REVIEW

**Comment by MrPetovan** [comment]
@rabuzarus Done.

File: include/text.php [hunk 1]
```diff
@@ -1271,6 +1268,22 @@ function prepare_body(&$item, $attach = false, $preview = false) {
...
+	if (!$is_preview && !($item['self'] && local_user() == $item['uid'])) {
+		if (!empty($item['content-warning']) && (!local_user() || !PConfig::get(local_user(), 'social', 'disable_cw', false))) {
+			$filter_reasons[] = L10n::t('Content warning: %s', $item['content-warning']);
+		}
+
+		$hook_data = [
+			'item' => $item,
+			'filter_reasons' => $filter_reasons
+		];
+		Addon::callHooks('content_filter', $hook_data);
```
Comment (annando) [inline include/text.php]: In the past we named the hooks after the function in which they had been. So "prepare_body_cw" or similar would be better, I think.

File: mod/settings.php [hunk 2]
```diff
@@ -213,6 +212,7 @@ function settings_post(App $a)
 		check_form_security_token_redirectOnErr('/settings/connectors', 'settings_connectors');
 
 		if (x($_POST, 'general-submit')) {
+			PConfig::set(local_user(), 'social', 'disable_cw', intval($_POST['disable_cw']));
```
Comment (annando) [inline mod/settings.php]: Concerning the category of the config value: Normally "system" is used for every core setting. Addon settings are using the addon name as category.

Comment (MrPetovan) [reply mod/settings.php]: A few lines below, there is a "ostatus" category, did it used to be an addon? What even is the purpose of the category?

File: include/text.php (hunk 1 above)
Comment (MrPetovan) [reply include/text.php]: Fair enough.

File: mod/settings.php (hunk 2 above)
Comment (annando) [reply mod/settings.php]: Yeah, we hadn't been totally strict with it. Just have a look at these settings:
```
PConfig::set(local_user(), 'system', 'ostatus_autofriend',intval($_POST['snautofollow']));
PConfig::set(local_user(), 'ostatus', 'default_group', $_POST['group-selection']);
PConfig::set(local_user(), 'ostatus', 'legacy_contact', $_POST['legacy_contact']);
```

Comment (MrPetovan) [reply mod/settings.php]: That's what I mean, why not use different category name then?

Comment (annando) [reply mod/settings.php]: This "ostatus" is the bad example. Problem is that there always could be a "social" addon. So possibly we had to use some special category names for addon settings?

Comment (annando) [reply mod/settings.php]: What do we want to do here?

Comment (MrPetovan) [reply mod/settings.php]: I'm planning on reverting to `system`.

Comment (annando) [reply mod/settings.php]: Okay.

**Comment by MrPetovan** [comment]
PR updated. I added a new behavior: when CW from Mastodon are prevented from collapsing the post content, the CW text is used as a title to keep its display within the post, except not as a reason to collapse the post.

It is necessary as people often misuse CW as part of their post, either for a joke or just as a prepend to the post body.

**Pull request merged**